identity-actor/src/actor/actor.rs

```diff
@@ -0,0 +1,466 @@
+// Copyright 2020-2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::any::Any;
+use std::collections::HashMap;
+use std::result::Result as StdResult;
+use std::sync::Arc;
+
+use crate::actor::errors::ErrorLocation;
+use crate::didcomm::message::DidCommPlaintextMessage;
+use crate::didcomm::termination::DidCommTermination;
+use crate::didcomm::thread_id::ThreadId;
+use crate::p2p::InboundRequest;
+use crate::p2p::NetCommander;
+use crate::p2p::RequestMessage;
+use crate::p2p::ThreadRequest;
+use crate::ActorConfig;
+use crate::ActorRequest;
+use crate::Asynchronous;
+use crate::Endpoint;
+use crate::Error;
+use crate::InvocationStrategy;
+use crate::RemoteSendError;
+use crate::RequestContext;
+use crate::RequestHandler;
+use crate::RequestMode;
+use crate::Result;
+use crate::SyncMode;
+use crate::Synchronous;
+
+use dashmap::DashMap;
+use futures::channel::oneshot;
+use identity_core::common::OneOrMany;
+use libp2p::Multiaddr;
+use libp2p::PeerId;
+use serde::de::DeserializeOwned;
+use uuid::Uuid;
+
+pub(crate) struct ActorState {
+  pub(crate) handlers: HandlerMap,
+  pub(crate) objects: ObjectMap,
+  pub(crate) threads_receiver: DashMap<ThreadId, oneshot::Receiver<ThreadRequest>>,
+  pub(crate) threads_sender: DashMap<ThreadId, oneshot::Sender<ThreadRequest>>,
+  pub(crate) peer_id: PeerId,
+  pub(crate) config: ActorConfig,
+}
+
+/// The [`Actor`] can be used to send and receive messages to and from other actors.
+///
+/// An actor is a frontend for an event loop running in the background, which invokes
+/// user-registered handlers and injects a copy of the actor into it. Actors can thus be cloned
+/// without cloning the event loop, and doing so is a relatively cheap operation.
+/// Handlers are registered at actor build time, using the [`ActorBuilder`](crate::ActorBuilder).
+///
+/// After shutting down the event loop of an actor using [`Actor::shutdown`], other clones of the
+/// actor will receive [`Error::Shutdown`] when attempting to interact with the event loop.
+#[derive(Clone)]
+pub struct Actor {
+  #[cfg(not(feature = "primitives"))]
+  pub(crate) commander: NetCommander,
+  #[cfg(feature = "primitives")]
+  pub commander: NetCommander,
```

Comment:
The type of `commander` seems to be the same regardless of the `primitives` feature being active or not, did you forget to change something during a refactor?

Reply:
It's only about visibility, enabling `primitives` users to access the commander directly for more low-level control. Regular users shouldn't really need to access the `NetCommander` so I left it crate-private.